### DISCUSSION TITLE
Multi-line "run" value valid or not?

### DISCUSSION THREAD

**Discussion by ondrejmirtes** [community/community#36656] [Actions]
I have a bit of a mystery here. Let's say my job has this step:

```
      - name: "Test"
        run: "${{ matrix.script }}"
```

And the matrix looks like this:

```
    strategy:
      matrix:
        include:
          - script: |
              cd e2e/result-cache-1
              echo -n > phpstan-baseline.neon
              ../../bin/phpstan -vvv
              patch -b src/Bar.php < patch-1.patch
              cat baseline-1.neon > phpstan-baseline.neon
              ../../bin/phpstan -vvv
              mv src/Bar.php.orig src/Bar.php
              echo -n > phpstan-baseline.neon
              ../../bin/phpstan -vvv
```

I know that multi-line scripts like these usually have `&& \` at the end of the lines to ensure the next command is called only if the previous one is successful. But it seems that in GitHub Actions it's not always necessary.

When I copy this script and run it locally on the repo, I get different result that seems like some of the lines were skipped during execution, or were executed in different order.

But also this script is kind of flaky and sometimes fails in GitHub Actions too.

I have two questions:

1) What should I do to have these multiline scripts executed correctly? Is it `&& \` at the end of each line or is something simpler possible?
2) What's the explanation for the current behaviour, what does Bash really do in this case?

Thank you.

**Comment by swissspidy** [comment]
Not an expert and maybe I am understanding something wrong, but here it's the `|` (pipe) in YAML ([docs](https://yaml.org/spec/1.2.2/#chapter-8-block-style-productions)) that allows you to write these multi-line strings.

In your example, 

```
          - script: |
              cd e2e/result-cache-1
              echo -n > phpstan-baseline.neon
```

Is equivalent to this

```
[
  "cd e2e/result-cache-1\n"
  "echo -n > phpstan-baseline.neon\n"
]
```

or in Bash script:

```
cd e2e/result-cache-1;
echo -n > phpstan-baseline.neon;
...;
```

in a Bash script, with each command executed separately, regardless of whether the previous command filed.

Whereas `a && b` is a way to chain commands in a way to only run `b` if `a` is successful (as you mention correctly). Doing that for multiple commands can result in very long lines, so 

```
cd e2e/result-cache-1 && \
echo -n > phpstan-baseline.neon
```

can be used to write that on multiple lines, but still use that same chaining.

If you don't want to write `&& \` on every line but still want the script to fail as soon as one command fails, you can use `set -e` at the beginning.

The `-e` tells Bash to stop immediately if one of the statements it executes has a non-zero exit code.

**Reply by ondrejmirtes** [reply]
Alright, so the flakiness and weird behaviour I thought was part of Bash syntax was actually a caching problem in my app 🤦 

And also - I have a suspicion that in GitHub Actions the `&& \` part isn't needed because `set -e` is already set by the system but I didn't verify this.
